### ISSUE TITLE
osiam-spring-boot-starter module

### ISSUE THREAD

**Issue by jvmlet** [osiam/osiam#121]
It would be nice to have osiam-spring-boot-starter module that allows to embed OSIAM as part of third-party Spring Boot application.
This approach will benefit from all spring-cloud-provided features (auto-Config/service-discovery/client side load balancing/etc.)
Thanks

**Comment by dacrome** [comment]
This [blog post](https://blog.frankel.ch/designing-your-own-spring-boot-starter-part-1) is a good starter I guess ;)

**Comment by dacrome** [comment]
I think two spring-boot-starter should be created:

`spring-boot-starter-osiam-server`
Like already described by @jvmlet, this is a starter which embeds OSIAM in an own backend service, maybe by providing `@EnableOsiamServer`. There could be other things like creating custom OSIAM beans e.g. to control which database will be used and for sure auto configuration. And yes, service discovery, maybe search for services which are osiam clients.

`spring-boot-starter-osiam-client`
This is a starter which can be used in services which needs access to OSIAM. I can think of `@EnableOsiamClient`. This starter is more a spring security module I guess. Provide default endpoint to be accessible by an OSIAM server, e.g. for service discovery or metrics. Provide annotations to secure spring mvc rest endpoints.

This is just a first draft, please throw in your ideas and comments.

**Comment by jvmlet** [comment]
According  to `spring-boot-starter-osiam-server`

> ... maybe search for services which are osiam clients.

Why server has to discover clients ? Isn't it should be the opposite way ?

According  to  `spring-boot-starter-osiam-client`

> Provide default endpoint to be accessible by an OSIAM server

The same question as above

> Provide annotations to secure spring mvc rest endpoints.

I think spring security already did the hard job, IMHO `spring-boot-starter-osiam-client` should be able to auto-register itself as spring security manager (implemented against discovered `spring-boot-starter-osiam-server`)

**Comment by dacrome** [comment]
> >    ... maybe search for services which are osiam clients.
> 
> Why server has to discover clients ? Isn't it should be the opposite way ?
> 
> >   Provide default endpoint to be accessible by an OSIAM server
> 
> The same question as above

Yes, you are absolutely right. The osiam clients should register themselves. So a default endpoint should be provided by the osiam backend service to make this possible.
